### PR TITLE
Refactor InstrumentationDefinitions generator

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR is a refactor of the InstrumentationDefinitions generator algorithm to use partial classes and to calculate the Wrapper Assembly fullname at runtime.

@DataDog/apm-dotnet

### CODE REVIEW

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1780 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/f4dcd6b09e206dc92e0d674b84baec93ee7a1dff) will **not change** line coverage
:heavy_check_mark: Merging #1780 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/f4dcd6b09e206dc92e0d674b84baec93ee7a1dff) will **not change** branch coverage
:no_entry: Merging #1780 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/f4dcd6b09e206dc92e0d674b84baec93ee7a1dff) will will **increase** complexity by `2`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/f4dcd6b09e206dc92e0d674b84baec93ee7a1dff) | #1780       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9265` / `14700` | `9327` / `14762` |          |
| Lines %   | `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4122` / `6410` | `4123` / `6410` |          |
| Branches %| `0.64%`      | `0.64%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7562`      | `7564`        |  `2`  :no_entry:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75004/artifacts?artifactName=Code%20Coverage%20Report_75004&api-version=6.0&%24format=zip)
* [For this PR #1780](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75020/artifacts?artifactName=Code%20Coverage%20Report_75020&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/f4dcd6b09e206dc92e0d674b84baec93ee7a1dff) | #1780       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.64%`      | `0.64%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7562`      | `7564`       |  `2` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Logging.LogRateLimiter](https://github.com/DataDog/dd-trace-dotnet/tree/b9ee96e73e95730f1124b50cc2b587a2fca3a8ab/tracer/src/Datadog.Trace/Logging/LogRateLimiter.cs) | `0.00%` :heavy_check_mark: | `-0.17%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.RuntimeMetrics.RuntimeMetricsWriter](https://github.com/DataDog/dd-trace-dotnet/tree/b9ee96e73e95730f1124b50cc2b587a2fca3a8ab/tracer/src/Datadog.Trace/RuntimeMetrics/RuntimeMetricsWriter.cs) | `0.00%` :heavy_check_mark: | `-0.08%` :no_entry:   | `0` :heavy_check_mark: |


The following classes were added in #1780:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.InstrumentationDefinitions](https://github.com/DataDog/dd-trace-dotnet/pull/1780/files) | `1.00%` | `1.00%` | `1` |


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75004/artifacts?artifactName=Code%20Coverage%20Report_75004&api-version=6.0&%24format=zip)
* [For this PR #1780](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75020/artifacts?artifactName=Code%20Coverage%20Report_75020&api-version=6.0&%24format=zip)

File: tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs [hunk 1]
```diff
@@ -595,7 +606,7 @@ public override bool Equals(object obj)
                 return Equals((CallTargetDefinitionSource)obj);
             }
 
-            public override int GetHashCode() => HashCode.Combine(TargetAssembly, TargetType, TargetMethod) + HashCode.Combine(TargetMinimumMajor, TargetMinimumMinor, TargetMinimumPatch,
+            public override int GetHashCode() => HashCode.Combine(IntegrationName, TargetAssembly, TargetType, TargetMethod) + HashCode.Combine(TargetMinimumMajor, TargetMinimumMinor, TargetMinimumPatch,
                                                                                                                                TargetMaximumMajor, TargetMaximumMinor, TargetMaximumPatch,
                                                                                                                                WrapperAssembly, WrapperType);
```
Comment (andrewlock) [inline tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs:611]: Just noticed, we shouldn't be adding two hash codes I think.
```suggestion
            public override int GetHashCode()
            {
                var hash = new HashCode();
                hash.Add(IntegrationName);
                hash.Add(TargetAssembly);
                hash.Add(TargetType);
                hash.Add(TargetMethod);
                hash.Add(TargetMinimumMajor);
                hash.Add(TargetMinimumMinor);
                hash.Add(TargetMinimumPatch);
                hash.Add(TargetMaximumMajor);
                hash.Add(TargetMaximumMinor);
                hash.Add(TargetMaximumPatch);
                hash.Add(WrapperAssembly);
                hash.Add(WrapperType);
                return hash.ToHashCode();
            }
```

**Review by andrewlock — APPROVED** [review]
Ironically, now we're not doing the filtering, it prob doesn't really need to be a partial class IMO 😅 But no big deal

File: tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs (hunk 1 above)
Comment (tonyredondo) [reply tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs:611]: Oh!, I'm pretty N00b with `HashCode`, Didn't know about that `Add` method.

Comment (andrewlock) [reply tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs]: Me neither until just now 😄 Only necessary when combining more than 8 values

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1780 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) will **not change** line coverage
:heavy_check_mark: Merging #1780 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) will **not change** branch coverage
:no_entry: Merging #1780 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) will will **increase** complexity by `2`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) | #1780       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9262` / `14709` | `9329` / `14771` |          |
| Lines %   | `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4122` / `6418` | `4125` / `6418` |          |
| Branches %| `0.64%`      | `0.64%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7570`      | `7572`        |  `2`  :no_entry:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75153/artifacts?artifactName=Code%20Coverage%20Report_75153&api-version=6.0&%24format=zip)
* [For this PR #1780](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75160/artifacts?artifactName=Code%20Coverage%20Report_75160&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) | #1780       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.64%`      | `0.64%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7570`      | `7572`       |  `2` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.Helpers.AsyncHelper](https://github.com/DataDog/dd-trace-dotnet/tree/d41b347d2bec6dc0ec54f6ee7a4526358fee4cb2/tracer/src/Datadog.Trace/ClrProfiler/Helpers/AsyncHelper.cs) | `0.00%` :heavy_check_mark: | `-0.08%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations.TaskContinuationGenerator`4](https://github.com/DataDog/dd-trace-dotnet/tree/d41b347d2bec6dc0ec54f6ee7a4526358fee4cb2/tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/TaskContinuationGenerator`1.cs) | `0.05%` :heavy_check_mark: | `0.09%` :heavy_check_mark:   | `0` :heavy_check_mark: |


The following classes were added in #1780:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.InstrumentationDefinitions](https://github.com/DataDog/dd-trace-dotnet/pull/1780/files) | `1.00%` | `1.00%` | `1` |


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75153/artifacts?artifactName=Code%20Coverage%20Report_75153&api-version=6.0&%24format=zip)
* [For this PR #1780](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75160/artifacts?artifactName=Code%20Coverage%20Report_75160&api-version=6.0&%24format=zip)

**Review by lucaspimentel — APPROVED** [review]

File: tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs (hunk 1 above)
Comment (lucaspimentel) [reply tracer/build/_build/PrepareRelease/GenerateIntegrationDefinitions.cs]: TIL! Looks like `Add` and `Combine` were added in .NET 5. R# does something like in older versions (using `Span` as example):
```csharp
int hashCode = TraceId.GetHashCode();
hashCode = (hashCode * 397) ^ SpanId.GetHashCode();
hashCode = (hashCode * 397) ^ OperationName.GetHashCode();
// ...same for each member that should participate in equality operations...
return hashCode;
```

**Pull request merged**